### DISCUSSION TITLE
Question about run submit OpenRadioss job with MPI on a machine with multiple nodes and cores

### DISCUSSION THREAD

**Discussion by li2qn** [OpenRadioss/OpenRadioss#1967] [Q&A]
Hello everyone,

I have a basic question about submitting job with MPI on a machine with multiple nodes and cores as well as multiple threads. Now I know how to run my job with MPI on a machine only has one node, but with multiple cores and threads. What is the command line to submit job with multiple nodes? Thanks

Regards
Lee

**Comment by MarianBulla** [comment]
Hello Lee,

Please see here: https://github.com/OpenRadioss/OpenRadioss/blob/main/INSTALL.md

It's explanined in detail.

![image](https://github.com/OpenRadioss/OpenRadioss/assets/105412502/979dcc10-45f2-41b2-bd9d-44ac47589af2)

Hope it helps you ?

Thanks and best regards,
Marian

**Comment by OlivierWienholtz** [comment]
Hi Lee, 

Just for clarification :
do you want  to run OpenRadioss on a single Node with Multiple cores or do you want to run on a cluster 
with Multiple Nodes (eg Machines)  connected toguether with a HighSpeed interconnect ? 

Thank You 
Best Regards,
Olivier

**Reply by li2qn** [reply]
Hey Oliver,

I want to run my job on a cluster with multiple nodes connected together. For example, I was trying to run my job with 6 nodes, each nodes has 32 cores. Thanks for your input

Regards
Lee

**Reply by OlivierWienholtz** [reply]
Thank You for the answer.

IT is true that the documentation on GitHub explains only Single node runs.
OpenRadioss runs on Clusters, but setting up is advanced stuff...

Indeed the way you are building / using OpenMPI depends on the Network cards, Drivers etc...
There are some things to take care of....

OpenMPI has a FAQ  at https://www.open-mpi.org/faq/?category=running
All command line arguments are explained in...

Here are some steps / I stay quite General as it may differ from one Cluster to another.

1. Understanding your Cluster : Nodes / Interconnect / Network Drivers
You have several nodes : 6 each node has 32 cores.
They are connected with a Network using Drivers.
You need to know how your Cluster is made to set correct flags to OpenMPI
Overall Job performance depends on this.

2. OpenMPI :
Build you will use.
You may have already an OpenMPI build dedicated to your Cluster, or the network drivers may be shipped with an OpenMPI build. If this is the case, if it is younger than 4.1.2 it will be usable.
Have a look to the Drivers documentation helps.

OpenMPI should be available on every node of your cluster & accessible from same PATH through all the nodes. Usually OpenMPI is on a shared directory. You should be able to execute ssh to different Cluster nodes w/o adding a password. Set this up with ssh keys.

3. Running OpenRadioss
OpenRadioss executable should be accessible from any node with same PATH. 
PATH/ LD_LIBRARY_PATH to OpenMPI must be set.

The easiest is that execution directory is accessible from any node. OpenRadioss running on the Cluster can access to Restart files etc...

Hostfile : 
You have a list of Hosts where you want to run your job.
To run on Multiple nodes : create a hostfile with this list in : 

Hostfile is either : 
1 line per MPI OpenRadioss process
1 line per use node with slots=[# Cores]

in your example : running on 2 nodes with 64 OpenRadioss SPMD process : the hostfile will look like : 
-------------------------------
node1 slots=32
node2 slots=32
--------------------------------
or 
------------------------------
node1
...                   32x times
node1
node2
....           32x times
node2
----------------------------------

Run OpenRadioss Starter with total number of MPI domains you want to use : (in my example it is 64)


starter_linux64_gf -np 64

Run OpenRadioss Engine : 

 (Path to OpenMPI/bin]/mpirun [Evt Network Settings]   --hostfile [hostfile] -np 64 engine_linux64_gf_ompi -i Engine_Deck

Hope it helps, 
Regards,
Olivier
